prog/expr.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2023 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package prog
+
+import (
+	"errors"
+	"fmt"
+)
+
+func (bo *BinaryExpression) Evaluate(finder ArgFinder) (uint64, bool) {
+	left, ok := bo.Left.Evaluate(finder)
+	if !ok {
+		return 0, false
+	}
+	right, ok := bo.Right.Evaluate(finder)
+	if !ok {
+		return 0, false
+	}
+	switch bo.Operator {
+	case OperatorCompareEq:
+		if left == right {
+			return 1, true
+		}
+		return 0, true
+	case OperatorCompareNeq:
+		if left != right {
+			return 1, true
+		}
+		return 0, true
+	case OperatorBinaryAnd:
+		return left & right, true
+	}
+	panic(fmt.Sprintf("unknown operator %q", bo.Operator))
+}
+
+func (v *Value) Evaluate(finder ArgFinder) (uint64, bool) {
+	if len(v.Path) == 0 {
+		return v.Value, true
+	}
+	found := finder(v.Path)
+	if found == SquashedArgFound {
+		// This is expectable.
+		return 0, false
+	}
+	if found == nil {
+		panic(fmt.Sprintf("no argument was found by %v", v.Path))
+	}
+	constArg, ok := found.(*ConstArg)
+	if !ok {
+		panic("value expressions must only rely on int fields")
+	}
+	return constArg.Val, true
+}
+
+func makeArgFinder(t *Target, c *Call, unionArg *UnionArg, parents ParentStack) ArgFinder {
+	return func(path []string) Arg {
+		f := t.findArg(unionArg.Option, path, nil, nil, parents, 0)
+		if f == nil {
+			return nil
+		}
+		if f.isAnyPtr {
+			return SquashedArgFound
+		}
+		return f.arg
+	}
+}
+
+func (r *randGen) patchConditionalFields(c *Call, s *state) (extra []*Call, changed bool) {
+	if r.inPatchConditional {
+		return nil, false
+	}
+	r.inPatchConditional = true
+	defer func() { r.inPatchConditional = false }()
+
+	var extraCalls []*Call
```

Comment:
If you gave names to return values, then you can just use them in the function and have an empty `return` in the end.
This avoids additional variable declarations and make it impossible to assign to a wrong variable (currently it's still possible to do `changed = true` which will be lost).